KAFKA-8172: Fix to close file handlers before renaming files/directories

kondetibharat opened the pull request:
Fix to close file handlers before renaming files / directories and open them back if required

Following are the file renaming scenarios:
- Files are renamed to .deleted so they can be deleted
- .cleaned files are renamed to .swap as part of Log.replaceSegments flow
- .swap files are renamed to original files as part of Log.replaceSegments flow

Following are the folder renaming scenarios:
- When a topic is marked for deletion, folder is renamed
- As part of replacing current logs with future logs in LogManager

In above scenarios, if file handles are not closed, we get file access violation exception

Idea is to close the logs and file segments before doing a rename and open them back up if required.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

kondetibharat committed:
Fix to close the handlers before renaming files and directories and open the handlers if required

Signed-off-by: REDMOND\bhkonde <bhkonde@microsoft.com>

hzxa21 commented:
@konradkalita Thanks for the patch. Looks good overall. I have three comments in general:

1. Seems like this problem only happens on Windows platform rather than Linux platform. Do you know about the specific reasons on why it happens only on Windows or why it doesn't happen on Linux? I am trying to see whether this patch is generic for all platforms and whether or not will introduce non-negligible overhead for other platforms.
2. If the close & reopen logic is only needed for the rename, I think it is more straight-forward to put the logic in the rename methods themselves instead of letting the caller to explicitly call close before and reopen after, which it is very easy for us to miss them. What do you think?
3. Is it possible to add a test case for the issue?

kondetibharat commented:
@hzxa21 Thanks for the code review.
1)	This consistently happens on Windows platform. I have seen few people complaining similar issues for some Linux flavors, but consensus is this is not an issue for most of the popular Linux flavors. This is a classic windows problem where if a handle is open for a file, NTFS does not allow it to be renamed / moved / deleted. All the file handles open on the file are required to be closed before renaming them. This is not much of an issue with Linux because of the way Linux file system is implemented. The way files are implemented in Linux, it allows to rename / move / delete files even if there are open handles to them.
The way I think about this patch is, before renaming any files / directories, we should always close the file channels. This is applicable to Linux as well. There are 2 fundamental scenarios. 
a)	Deletion of directories since the topic is deleted and deletion of segment files by cleaner thread. We believe, we should close the log file in case of log deletion and close the segment in case of segment file cleanup. Since no one should be reading these files, it is safe to close.
b)	 During swapping flow, where ‘.cleaned’ files are renamed to ‘.swap’ files and then they are renamed to actual log files. Here also we believe segment should be closed during renaming and re-opened after the swap.
Overall gist is, files should be closed before renaming them and should be opened back if required. This should be applicable to Linux as well
2)	Yes, initially we implemented the change in rename files method. But that change turned out to be 
ugly. Here is the abandoned pull request done such way.
https://github.com/Microsoft/kafka/pull/17/files
Problem was, there was no need to re-open ‘.deleted’ files. So we had put an if check to not to re-open ‘.deleted’ file. Also, it resulted in changes at many places.
Also, I felt like it was not a cleaner fix where we are not solving the root problem. Which is to close the segment before they are renamed / deleted. Also, we wanted this to be a generic fix, rather than specific to Windows.

3)	Yes, I can add some test cases. We wanted to get a initial feedback on the idea before adding tests.

raj1003 commented:
> @hzxa21 Thanks for the code review.
> 
> 1. This consistently happens on Windows platform. I have seen few people complaining similar issues for some Linux flavors, but consensus is this is not an issue for most of the popular Linux flavors. This is a classic windows problem where if a handle is open for a file, NTFS does not allow it to be renamed / moved / deleted. All the file handles open on the file are required to be closed before renaming them. This is not much of an issue with Linux because of the way Linux file system is implemented. The way files are implemented in Linux, it allows to rename / move / delete files even if there are open handles to them.
>    The way I think about this patch is, before renaming any files / directories, we should always close the file channels. This is applicable to Linux as well. There are 2 fundamental scenarios.
>    a)	Deletion of directories since the topic is deleted and deletion of segment files by cleaner thread. We believe, we should close the log file in case of log deletion and close the segment in case of segment file cleanup. Since no one should be reading these files, it is safe to close.
>    b)	 During swapping flow, where ‘.cleaned’ files are renamed to ‘.swap’ files and then they are renamed to actual log files. Here also we believe segment should be closed during renaming and re-opened after the swap.
>    Overall gist is, files should be closed before renaming them and should be opened back if required. This should be applicable to Linux as well
> 2. Yes, initially we implemented the change in rename files method. But that change turned out to be
>    ugly. Here is the abandoned pull request done such way.
>    https://github.com/Microsoft/kafka/pull/17/files
>    Problem was, there was no need to re-open ‘.deleted’ files. So we had put an if check to not to re-open ‘.deleted’ file. Also, it resulted in changes at many places.
>    Also, I felt like it was not a cleaner fix where we are not solving the root problem. Which is to close the segment before they are renamed / deleted. Also, we wanted this to be a generic fix, rather than specific to Windows.
> 3. Yes, I can add some test cases. We wanted to get a initial feedback on the idea before adding tests.

We are backing the patch in our test environment.

gladhiola commented:
Dear All,

Could you please help me how could I apply this patchset? I would like to use it in a maven and in a gradle based java project also, and in both cases the kafka package exists as a dependency.

bastiao commented:
Any plans to bring this to upstream?
